### PR TITLE
Introduce doIntegrateHierarchy() virtual function.

### LINKED ISSUE #1636: executeIntegrateHierarchyCallbackFcns behavior in the outermost derived class
Hi @boyceg ,

We have ```PhaseChangeHierarchyIntergrator (PCHI)``` derived from ```AdvDiffSemiImplicitHierarchyIntegrator (ADSIHI)``` class. The derived class PCHI calls the ```ADSIHI::integrateHierarchy()``` in its ```integrateHierarchy()```. The derived class PCHI also calls the ```executeIntegrateHierarchyCallbackFcns()``` at the end of ```integrateHierarchy()```.  This is needed if we want to modify or reset the solution before the next cycle. The problem is the base class ADSIHI also calls the ```executeIntegrateHierarchyCallbackFcns``` function in the ```integrateHierarchy``` function. Eventually, ```executeIntegrateHierarchyCallbackFcns``` is executed twice: One in the ADSIHI and the other in the PCHI. Do you have any suggestions for this?

### PR DESCRIPTION
**Pull request by abarret**
Fixes #1636. This introduces a new protected pure virtual function `doIntegrateHierarchy()` in `HierarchyIntegrator`. Previously, `integrateHierarchy()` was designed to do two things: (1) perform an integration cycle and (2) execute post integration callbacks. The design principle was that the class furthest down the hierarchy ends up calling the post integration callbacks. This causes problems (e.g. in the phase change code from @amneetb and @Raamkrishnan) where child classes want to call the parent class's `integrateHierarchy()` to perform action (1) but not (2).

This changes the design principle so that the parent class `HierarchyIntegrator::integrateHierarchy()` executes post integration callbacks after `doIntegrateHierarchy()` is called. Child classes no longer need to execute the callbacks. I've left `integrateHierarchy()` as a virtual function in case some future application needs to modify the implementation, although I don't think this is necessary.

@boyceg This is the kind of change I was suggesting. This clearly delineates the two purposes of `integrateHierarchy()`. The function `doIntegrateHierarchy()` should _only_ integrate equations.

### IBAMR Pull Request Checklist
- [x] Does the test suite pass?
- [x] Was clang-format run?
- [x] Were relevant issues cited? Please remember to add `Fixes #12345` to close
      the issue automatically if we are fixing the problem.
- [x] Is this a change others will want to know about? If so, then has a
      changelog entry been added?
- [x] If new data structures have been added to a class then have they been set
      up appropriately for restarts? If so, ensure that the restart version
      number is incremented.
- [x] Does this change include a bug fix or new functionality? If so, a new test
      or tests should be added. New tests should run quickly (less than a minute
      in release mode). If possible, an older test should gain a new option so
      that we do not need to compile more test executables.
- [x] Did you (if your account has permission to do so) set relevant labels on
      GitHub for the pull request?

### CODE REVIEW

**Comment by drwells** [comment]
> This changes the design principle so that the parent class HierarchyIntegrator::integrateHierarchy() executes post integration callbacks after doIntegrateHierarchy() is called. Child classes no longer need to execute the callbacks. I've left integrateHierarchy() as a virtual function in case some future application needs to modify the implementation, although I don't think this is necessary.

That's definitely the right thing to do. My main concern with this patch is that it breaks basically any every non-IBAMR class which inherits from `HierarchyIntegrator`.

Perhaps we can do things in a more compatible way: I'll take a quick look and report back.

**Comment by abarret** [comment]
> That's definitely the right thing to do. My main concern with this patch is that it breaks basically any every non-IBAMR class which inherits from `HierarchyIntegrator`.

Changing which class is responsible for executing the callback function will break a lot of classes that inherit from `HierarchyIntegrator` anyway. The only real change required for downstream classes is changing the name `integrateHierarchy()` to `doIntegrateHierarchy()`, so the change is minimal. 

We could provide a no-op implementation of `HierarchyIntegrator::doIntegrateHierarchy()` so that classes that want to do everything in `integrateHierarchy()` still can.

**Comment by abarret** [comment]
Another question is should we do the same thing for all the callback functions: `postprocessIntegrateHierarchy`, `preprocessIntegrateHierarchy`, `applyGradientDetector`, and `regridHierarchy`.

**Comment by drwells** [comment]
I'd rather not modify those extension mechanisms (i.e,. we should go outside-in and then inside-out for preprocessing and postprocessing) if we don't have to.

**Comment by drwells** [comment]
How about we rename `doIntegrateHierarchy()` to `integrateHierarchySpecialized()` to match the naming convention? Similarly, I think `integrateHierarchy()` should not be virtual: people should use the documented extension mechanisms instead of making this even more complex (and we're breaking stuff now in an incompatible way).  Other than that (and the CI errors) I think this is fine.

My basic attempt (move `executeIntegrateHierarchyCallbackFcns()` into `HierarchyIntegrator::integrateHierarchy()` didn't work (I'm not sure why) so I don't think there's another straightforward way to fix this. We should also check that derived classes call the parent class `integrateHierarchySpecialized()` at the top of their own version of that function.

File: ibtk/include/ibtk/HierarchyIntegrator.h [hunk 1]
```diff
@@ -409,14 +409,15 @@ class HierarchyIntegrator : public SAMRAI::mesh::StandardTagAndInitStrategy<NDIM
     virtual void preprocessIntegrateHierarchy(double current_time, double new_time, int num_cycles = 1);
 
     /*!
-     * Pure virtual method to advance data from current_time to new_time.
+     * Virtual method to advance data from current_time to new_time. The current implementation calls
```
Comment (drwells) [inline ibtk/include/ibtk/HierarchyIntegrator.h:412]:
```suggestion
     * Advance data from current_time to new_time. The current implementation calls
```

File: ibtk/include/ibtk/HierarchyIntegrator.h [hunk 2]
```diff
@@ -697,6 +698,16 @@ class HierarchyIntegrator : public SAMRAI::mesh::StandardTagAndInitStrategy<NDIM
     void putToDatabase(SAMRAI::tbox::Pointer<SAMRAI::tbox::Database> db) override;
 
 protected:
+    /*!
+     * Pure virtual function that integrates the hierarchy from current_time to new_time.
+     *
+     * Implementations of this virtual function are not required to synchronize
+     * data on the patch hierarchy.  Data synchronization may be done
+     * (optionally) in a specialization of the public virtual member function
+     * postprocessIntegrateHierarchy().
```
Comment (drwells) [inline ibtk/include/ibtk/HierarchyIntegrator.h:707]: Could you add a `@note` here explaining why we did this, i.e., some discussion of when callbacks are called?

Comment (drwells) [reply ibtk/include/ibtk/HierarchyIntegrator.h:707]: Could you also mention that derived classes should typically call the parent class `integrateHierarchySpecialized()` at the top of their own implementation of this function?

File: src/IB/IBExplicitHierarchyIntegrator.cpp [hunk 3]
```diff
@@ -156,9 +156,10 @@ IBExplicitHierarchyIntegrator::preprocessIntegrateHierarchy(const double current
 } // preprocessIntegrateHierarchy
 
 void
-IBExplicitHierarchyIntegrator::integrateHierarchy(const double current_time, const double new_time, const int cycle_num)
+IBExplicitHierarchyIntegrator::integrateHierarchySpecialized(const double current_time,
+                                                             const double new_time,
+                                                             const int cycle_num)
 {
-    IBHierarchyIntegrator::integrateHierarchy(current_time, new_time, cycle_num);
```
Comment (drwells) [inline src/IB/IBExplicitHierarchyIntegrator.cpp:161]: I'm actually surprised that this works. We should definitely still call `IBHierarchyIntegrator::integrateHierarchySpecialized()` here even if that function doesn't do anything.

**Review by drwells — CHANGES_REQUESTED** [review]
Thanks - I have a few more minor suggestions.

I think that it's important that we keep the idiom 'outside-in and then inside-out' working for hierarchy integrators. We use that (or at least we should - I think I fixed all of them) for preprocessing and postprocessing too so that derived classes don't mess around with parent class data.

**Comment by drwells** [comment]
@abarret can you get this done today? If not lets leave it for 0.15.

File: src/IB/IBExplicitHierarchyIntegrator.cpp (hunk 3 above)
Comment (abarret) [reply src/IB/IBExplicitHierarchyIntegrator.cpp:161]: I added an empty implementation of `HierarchyIntegrator::integrateHierarchySpecialized()` so we can make this call.

Note the only integrator that doesn't call it's immediate base class is `BrinkmanAdvDiffSemiImplicitHierarchyIntegrator`, which inherits from `AdvDiffSemiImplicitHierarchyIntegrator` but calls `AdvDiffHierarchyIntegrator::integrateHierarchySpecialized()`. This is consistent with the class's behavior prior to this PR.

**Comment by abarret** [comment]
> @abarret can you get this done today? If not lets leave it for 0.15.

I've made the requested changes and squashed the commits.

**Review by drwells — APPROVED** [review]

**Pull request merged**